### DISCUSSION TITLE
APU_MEM reverse engineering.

### DISCUSSION THREAD

**Discussion by flukejones** [seerge/g-helper#3148] [General]
Hi Sergei, wondering if you can help me here - How did you discover the `APU_MEM` https://github.com/seerge/g-helper/commit/32c4f6168c97a589a3ccc0902d020c46f4cb9655 setting?

I've been trawling through the various files of windows trying to match something but apparently coming up blank.

**Comment by seerge** [comment]
@flukejones hello. 

I just changed APU setting on Rog Ally and did endpoint scan after each change. ``0x000600C1`` was returning correct APU memory each time

**Reply by flukejones** [reply]
I see. Would you mind sharing with me the method you used for scan? Seems like I've been doing things the hard way :|

**Reply by seerge** [reply]
There is a powershell scan-script posted here https://github.com/seerge/g-helper/issues/38

Also app by itself can do it as well https://github.com/seerge/g-helper/blob/ab08bd8ab45774b3e1961ddff69b6281a2f6e733/app/AsusACPI.cs#L759

**Reply by flukejones** [reply]
Very cool! Thanks for all your work on the Windows side.

I do wish i could find the DLL that manages the GPU mem size though, it's elusive so far. Main reason is because I want to snoop through a few things as part of a verification for myself. You wouldn't happen to know the dll name?

**Reply by seerge** [reply]
It's not a DLL i think, it's BIOS / firmware that handles this same as pretty much any settings that you set via all this endpoints

**Reply by flukejones** [reply]
Yeah that's true but so far most of those have a function in a DLL calling a method such as `One_WMIMethod_INT`. An example is `GetGPUModeStatus`. I've mapped most, and I think the `Cutter` app is available on Windows? Never looked.

https://gitlab.com/asus-linux/reverse-engineering/-/blob/master/WMI_SPEC.md?ref_type=heads

While I've got your attention do you happen to also know of a HID method to get the MCU version?

**Reply by seerge** [reply]
@flukejones ah, for this specific setting it should be some part of Armoury Crate package for Ally. But I honestly have no idea, as I didn't check it that way.

I only spotted endpoint that is responsible for both setting and reporting current APU mem status here.

And I'm not aware about HID methods to check MCU version. Do you want it to check if a firmware update is needed ?

**Reply by flukejones** [reply]
Yeah I've been working with asus to fix a long standing issue in Linux for the Ally devices. Took a *lot* of work to really narrow down the exact cause, asus even set up a full Linux dev environment to help - so there is a new FW coming, and I'm trying to find a good way to figure the version.

Guess I'll just have to ask them :grin: Hmm actually, I'll check a wireshark capture when i open the armoury versions tab.

**Reply by seerge** [reply]
@flukejones are you sure that they use USB to check for MCU version in a first place ? Could be just another WMI method or something less obvious ... :) I might borrow it too tho, cause in Updates section G-Helper lists firmware update for Ally, but i can't check actual version of course

**Reply by flukejones** [reply]
At this point no idea, but the updater logs a lot of handy info `[Initialize] Get FW Version : FGA80100.RC71LS.312` for example. So I've been searching on that.

If I find it I'll let you know for sure.

**Reply by flukejones** [reply]
Hey mate I found the HID call for MCU version: See the commit at https://github.com/flukejones/linux/commit/095fe46b03244aeb3600131e68945f8d1a57c1af

Request (endpoint 0x83 or 0x87):
```
Ally 1 MCU1, Ally X:
5a,05,03,31,00,20
Ally 1 MCU2:
5a,05,04,31,00,20
```

Response:
```
5a050331001a1346474138303130302e524337314c532e333038000000000000

0000   1c 00 50 b0 0c 67 8f a0 ff ff 00 00 00 00 08 00   ..P..g..........
0010   01 01 00 03 00 80 02 20 00 00 00 03 5a 05 03 31   ....... ....Z..1
0020   00 1a 13 46 47 41 38 30 31 30 30 2e 52 43 37 31   ...FGA80100.RC71
0030   4c 53 2e 33 30 38 00 00 00 00 00 00               LS.308......
```

Hope this is helpul for you

**Reply by seerge** [reply]
Oh, so there are 2 different MCUs ? What does second MCU request (`04-31-00-20`) respond as a version ? Same thing ?

**Reply by flukejones** [reply]
The second one is the same except RC71LM. I'm actually not certain which is which, I'll check the log

The Ally X looks like this:
```
Sep 23 12:35:15 steamfork kernel: Buffer content (hex):
Sep 23 12:35:15 steamfork kernel: 5a 05 03 31 00 1a 17 46 47 41 38 30 31 30 30 2e
Sep 23 12:35:15 steamfork kernel: 52 43 37 32 4c 41 2e 33 31 32 5f 54 30 31 00 00
Sep 23 12:35:15 steamfork kernel: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 23 12:35:15 steamfork kernel: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Sep 23 12:35:15 steamfork kernel:
Sep 23 12:35:15 steamfork kernel: asus_rog_ally 0003:0B05:1B4C.0004: Parsed string: RC72LA
Sep 23 12:35:15 steamfork kernel: asus_rog_ally 0003:0B05:1B4C.0004: Parsed int: 312
```

**Reply by flukejones** [reply]
Logs:
```
[Initialize] Productname :RC71L
[Initialize] Second Productname :RC71LS
[Initial] fwfiles : System.String[]
[Initial] fwfiles.Length : 2
[Initial] fwfiles[i] : FGA80100.RC71LM.315.bin
[Initial] keyboardtype : 3
[Initial] RC71LS_Name_0 : RC71LM
[Initial] RC71LS_Name_0 : RC71LS
[Initial] fwfiles[i] : FGA80100.RC71LS.315.bin
...
[Initialize] ****************** Main IC update ******************
[Backup] Backup Main FW version = FGA80100.RC71LM.312
[Initialize] Get Main FW version : FGA80100.RC71LM.312
```

So M is `main`, S is `secondary`

**Reply by seerge** [reply]
@flukejones ok, thanks! But as I understand firmware updated updates both (and version is the same), so it's worth to check only one probably

**Reply by flukejones** [reply]
Indeed correct. The updater app for the Ally 1 contains the bin files for each MCU. They are always done as a pair and can't be mismatched.
